org/mozilla/jss/util/NativeEnclosure.java

```diff
@@ -0,0 +1,42 @@
+package org.mozilla.jss.util;
+
+/**
+ * Classes extending NativeEnclsoure wrap a single NativeProxy instance,
+ * allowing it to be accessed from the JNI layer but be allocated and scoped
+ * from the Java layer.
+ *
+ * This allows us to data parameters in Java and call open(...) when all
+ * necessary variables have been created. At the end of this object's use,
+ * close must be called to free allocated memory.
+ *
+ * Because this class implements AutoCloseable, it is suggested to add
+ * constructors to derived classes which call open; this'll allow a single
+ * try-with-resources block to scope the lifetime of this object:
+ *
+ * <pre>
+ * try (NEC obj = new NEC(...)) {
+ *      // ... do something with obj ...
+ * }
+ * </pre>
+ */
+public abstract class NativeEnclosure implements AutoCloseable {
+    public NativeProxy ptr;
+    public long ptr_size;
+
+    public final void open() throws Exception {
+        acquireNativeResources();
```

Comment:
Any concern about concurrency? Do we need some kind of locking mechanism?

Reply:
Wouldn't these objects go inside the lock? Not really sure what the concurrency model would be here anyways. At least, with our KDF, concurrency shouldn't be an issue since it'll all be set up serially (and the actual call looks like `open(...) / call_jni_kdf() / close(...)` deep within the inner details of JSS implementation.

This `NativeEnclosure` isn't really a public-facing abstract class; it is mroe of an implementation detail of how we do stuff.

Reply:
Actually, maybe not concurrency. Is open() idempotent as well? Is there any possibility that `open()` is called multiple times on the same object?

Reply:
We could do:

```java
public final void open() throws Exception {
    if (ptr == null) {
        acquireNativeResources();
    }
}
```

And do similar for `close()` if that'd be preferable. Then if the user wishes to call `open()` multiple times, they'd have to call `close()` first, thus freeing the memory.

Reply:
OK